### DISCUSSION TITLE
Device control plane event configuration against EdgeX v2.3

### DISCUSSION THREAD

**Discussion by NileshDixit-eaton** [edgexfoundry/community#290] [Q&A] [answered]
Hi Team,

We are currently on EdgeX v2.3 and trying to test device basic CRUD operation using meta-data control plane event feature
Tried to change metadata configuration.toml to have 
~ edgex/core/2.0/core-metadata/Notifications/PostDeviceChanges set to True 
edgex/core/2.0/core-metadata/MessageQueue/PublishTopicPrefix set to 'edgex/system-events'
and other dependent Notification configuration as well
Also kept mosquito listening to topic edgex/system-events/#
But with above configuration change, I see metadata service core-dump with below reason

`
Jul 03 03:47:32 BL-Edge-Gateway-42-83-37 core-metadata[2701]: level=WARN ts=2024-07-03T07:47:32.820126686Z app=core-metadata source=device.go:239 msg="unable to publish Device System Event: MessageBus Client not available. Please update RequireMessageBus and MessageQueue configuration to enable sending System Events via the EdgeX MessageBus"
Jul 03 03:47:32 BL-Edge-Gateway-42-83-37 core-metadata[2701]: panic: runtime error: invalid memory address or nil pointer dereference
Jul 03 03:47:32 BL-Edge-Gateway-42-83-37 core-metadata[2701]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0xdbfcf4]
Jul 03 03:47:32 BL-Edge-Gateway-42-83-37 core-metadata[2701]: goroutine 163 [running]:
Jul 03 03:47:32 BL-Edge-Gateway-42-83-37 core-metadata[2701]: github.com/edgexfoundry/edgex-go/internal/core/metadata/application.sendNotification({0x10a9150, 0x19f3900}, 0x189fe60, {0x1c87f50, 0x13}, {0xe1c208, 0xf})
Jul 03 03:47:32 BL-Edge-Gateway-42-83-37 core-metadata[2701]:         github.com/edgexfoundry/edgex-go/internal/core/metadata/application/notify.go:265 +0x370
Jul 03 03:47:32 BL-Edge-Gateway-42-83-37 core-metadata[2701]: created by github.com/edgexfoundry/edgex-go/internal/core/metadata/application.addDeviceCallback in goroutine 145
Jul 03 03:47:32 BL-Edge-Gateway-42-83-37 core-metadata[2701]:         github.com/edgexfoundry/edgex-go/internal/core/metadata/application/notify.go:60 +0x460
Jul 03 03:47:32 BL-Edge-Gateway-42-83-37 systemd[1]: edgex-core-metadata.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Jul 03 03:47:32 BL-Edge-Gateway-42-83-37 systemd[1]: edgex-core-metadata.service: Failed with result 'exit-code'.
`
**Query : Is some more configuration missing to get device CRUD update event from metadata service**
Attached metadata service configuration.toml

cc @cloudxxx8 
[metadata-configuration_toml.txt](https://github.com/user-attachments/files/16078916/metadata-configuration_toml.txt)

**Comment by cloudxxx8** [comment]
```
# Removing support-notifications client to make it severable since core-metadata is required
# and support-notifications is not. Uncomment to use support-notifications to notify about
# device changes.

[Clients]
#  [Clients.support-notifications]
#  Protocol = 'http'
#  Host = '127.0.0.1'
#  Port = 59860
```

**Reply by cloudxxx8** [reply]
You missed the notifciation client setting when you enable the notification

**Reply by NileshDixit-eaton** [reply]
Hi @cloudxxx8 
Thanks for prompt response
I did enabled the notification client but I still see the below logs on CRUD operation

`Jul 03 06:56:07 BL-Edge-Gateway-42-83-37 core-metadata[2300]: level=WARN ts=2024-07-03T10:56:07.579866124Z app=core-metadata source=device.go:239 msg="unable to publish Device System Event: MessageBus Client not available. Please update RequireMessageBus and MessageQueue configuration to enable sending System Events via the EdgeX MessageBus"
`
Anything more missing?

**Reply by cloudxxx8** [reply]
The MessageQueue config is not recognized by the service.  Your TOML format may be wrong
```
# We need a MessageQueue configuration so that if we are using Keeper for
# a config provider, we can see config changes in realtime
[MessageQueue]

# Configuration for mqtt message bus
Protocol = 'tcp'
Host = '127.0.0.1'
Port = 59000
Type = 'mqtt'
AuthMode = 'none'  # required for redis messagebus (secure or insecure)
SecretName = "mqtt-bus"
PublishTopicPrefix = 'edgex/system-event' # /<device-profile-name>/<device-name>will be added to this Publish Topic prefix
SubscribeEnabled = true  # a lie, go-mod-configuration will subscribe if using Keeper
SubscribeTopic = "edgex/events/device/#"  # required for subscribing to Events from MessageBus
    # MQTT Specific options
    [MessageQueue.Optional]
    ClientId = "core-metadata"
    Qos            = "1" # Quality of Service values are 0 (At most once), 1 (At least once) or 2 (Exactly once)
    KeepAlive      = "10" # Seconds (must be 2 or greater)
    Retained       = "false"
    AutoReconnect  = "true"
    ConnectTimeout = "5" # Seconds
    SkipCertVerify = "true"
```

Please try to eliminate the empty line after `[MessageQueue]`:
```
# We need a MessageQueue configuration so that if we are using Keeper for
# a config provider, we can see config changes in realtime
[MessageQueue]
# Configuration for mqtt message bus
Protocol = 'tcp'
Host = '127.0.0.1'
```

**Reply by NileshDixit-eaton** [reply]
@cloudxxx8 
But the MessageQueue configuration under keeper are still valid irrespective of that empty line added or removed
keeper /kvs/key/{key} GET against key edgex/core/2.0/core-metadata/MessageQueue
`{
  "apiVersion": "v2",
  "statusCode": 200,
  "response": [
    {
      "key": "edgex/core/2.0/core-metadata/MessageQueue/Optional/AutoReconnect",
      "created": 1720013317073,
      "modified": 1720013317073,
      "value": "true"
    },
    {
      "key": "edgex/core/2.0/core-metadata/MessageQueue/Optional/SkipCertVerify",
      "created": 1720013317073,
      "modified": 1720013317073,
      "value": "true"
    },
    {
      "key": "edgex/core/2.0/core-metadata/MessageQueue/Optional/Retained",
      "created": 1720013317073,
      "modified": 1720013317073,
      "value": "false"
    },
    {
      "key": "edgex/core/2.0/core-metadata/MessageQueue/Optional/Qos",
      "created": 1720013317073,
      "modified": 1720013317073,
      "value": "1"
    },
    {
      "key": "edgex/core/2.0/core-metadata/MessageQueue/Optional/ClientId",
      "created": 1720013317074,
      "modified": 1720013317074,
      "value": "core-metadata"
    },
    {
      "key": "edgex/core/2.0/core-metadata/MessageQueue/Optional/ConnectTimeout",
      "created": 1720013317072,
      "modified": 1720013317072,
      "value": "5"
    },
    {
      "key": "edgex/core/2.0/core-metadata/MessageQueue/Optional/KeepAlive",
      "created": 1720013317072,
      "modified": 1720013317072,
      "value": "10"
    },
    {
      "key": "edgex/core/2.0/core-metadata/MessageQueue/Host",
      "created": 1720013317075,
      "modified": 1720013317075,
      "value": "127.0.0.1"
    },
    {
      "key": "edgex/core/2.0/core-metadata/MessageQueue/SubscribeEnabled",
      "created": 1720013317076,
      "modified": 1720013317076,
      "value": "true"
    },
    {
      "key": "edgex/core/2.0/core-metadata/MessageQueue/Protocol",
      "created": 1720013317076,
      "modified": 1720013317076,
      "value": "tcp"
    },
    {
      "key": "edgex/core/2.0/core-metadata/MessageQueue/Port",
      "created": 1720013317069,
      "modified": 1720013317069,
      "value": "59000"
    },
    {
      "key": "edgex/core/2.0/core-metadata/MessageQueue/SecretName",
      "created": 1720013317069,
      "modified": 1720013317069,
      "value": "mqtt-bus"
    },
    {
      "key": "edgex/core/2.0/core-metadata/MessageQueue/Type",
      "created": 1720013317075,
      "modified": 1720013317075,
      "value": "mqtt"
    },
    {
      "key": "edgex/core/2.0/core-metadata/MessageQueue/PublishTopicPrefix",
      "created": 1720013317075,
      "modified": 1720013317075,
      "value": "edgex/system-events"
    },
    {
      "key": "edgex/core/2.0/core-metadata/MessageQueue/SubscribeTopic",
      "created": 1720013317075,
      "modified": 1720013317075,
      "value": "edgex/events/device/#"
    },
    {
      "key": "edgex/core/2.0/core-metadata/MessageQueue/AuthMode",
      "created": 1720013317069,
      "modified": 1720013317069,
      "value": "none"
    }
  ]
}`

**Comment by cloudxxx8** [comment, marked as answer]
sorry, I missed one statement from the error log
 Please update **RequireMessageBus** and MessageQueue configuration
https://github.com/edgexfoundry/edgex-go/blob/4f1b2626cb078c7149dc655301029b2517b4b500/cmd/core-metadata/res/configuration.toml#L1-L4

You need to add
`RequireMessageBus = true`
to the configuration

**Reply by NileshDixit-eaton** [reply]
yes, I was thinking of same but generally we put this under [Service] configuration and with metadata config.toml not having [Service] configuration so was not sure, anyway will try this and let you know Thanks
